Issue# 155 Entity Stuff

jdisser opened the pull request:
Not sure about the procedure to rewrite the abstract class to clean up the partially unused methods. Also a glitch in the tests that maven run on the build.

jdisser commented:
After a bit of digging I see that the class was supposed to implement Comparable. So @Override might have been added to the compareTo method as a precaution but it's now unnecessary. I removed the implementation of Comparable and the @Override annotations related to that method. Some confusion still remains. Do I need to implement the Java Comparable interface for those entities that need it, like Questions?? Or does simply adding a compareTo method handle that without implementing the interface?  

Still have some basic questions on how to test these changes in general and confirming that I'm selecting methods for removal in a sane manner.

jdisser commented:
Added the Comparable interface to Questions to the effect. No effect on the test results when building BUT I am getting an error 
`Could not find a value for QUIZKI_JDBC_URL in System.getProperty()`
that I can't figure out how to resolve so I don't think my test results are valid. This version works when taking an exam on the site. 
Ideally I would like to switch between these last two commits and see it break in the tests if it should.

jdisser commented:
Thanks! Got it on the comments, that didn't 'smell' right when I did it should go with that next time. Will do on the discussion early next week. 
I'm working on the id's at the moment and seeing about moving that functionality into AbstractEntity, along with the default value. I should have another commit by the time we talk on that too. I need to understand better why AbstractEntity is only inherited by 5 Entities and if this makes sense given that situation.

jdisser commented:
Reverted the previous commit that changed `Exam` to an AbstractTextEntity. This forced it to have an unused `text` field and there was no straightforward way to prevent that.

jdisser commented:
@haxwell This pull request is ready for review!

haxwell commented:
Hey Jim.. Thanks! This looks like a big one! I haven't got a chance to go all the way through it.. Can you  tell me, though, what was your plan of attack when removing the ID field on some entities (and not others).. The original purpose of those was to create a new ID when new objects were persisted, and to store the existing ID on already-persisted objects.

jdisser commented:
Hey Jonathan I was thinking this would be a lot to review. I moved the id field into AbstractEntity which is the root class for all entities. The `@MappedSuperclass` annotation causes the id field to be persisted as an ID in the tables created for all the subclasses, so this has not changed, all entities still have an ID thru inheritance. The annotation also prevents a table being created for the Abstract class which keeps our database schema intact. That was my objective with these changes to not alter the database.  

There is also a new Abstract class called AbstractTextEntity which extends AbstractEntity. It adds a `text` field and it's setters and getters. It's also a `@MappedSuperclass` and will generate these fields in it's children's tables. Some entities extend AbstractTextEntity and so have both an `id` and a `text` field in the database. 

There was an interesting problem with `Exam` since it needed a `getText` method but did not have a `text` field. I had to put an extra `getText` method in AbstractEntity to handle this. I tried to use the AbstractTextEntity and it compiled and ran but did not return `Exams` from the database. This also added a `text` field to Exam that would have been persisted and need a change to the database that would never be used. I realized all of this after a commit so I reverted that change and added the "unclean" `getText` method to AbstractEntity and went from there.

haxwell commented on src/main/java/com/haxwell/apps/questions/entities/Difficulty.java:
```diff
@@ -23,10 +23,13 @@
 import javax.persistence.Table;
 
+//This class did not implement javax.persistance.GenerationType using it now with ATE might cause conflicts in database
+
```
Yes, that is because we do not generate new Difficulty objects. There is a defined set of them, and we only use those. Question, Choice, Exam, Vote, etc.. we generate new ones of those.

haxwell commented on src/main/java/com/haxwell/apps/questions/utils/DifficultyUtil.java:
```diff
@@ -29,42 +30,33 @@ public static int convertToInt(Difficulty d) {
+			if (parameter.toLowerCase().equals(de.getValString().toLowerCase()))
+				return (int)de.getRank();
+		}
+		return 1; //this is an accepted value? Should it be 0 or -1 or a default??
```
Yes, the '1' is correct. We want to indicate that the Difficulty is Junior by default. So if `parameter` is not a match, we return Junior

haxwell commented on src/main/java/com/haxwell/apps/questions/entities/AbstractEntity.java:
```diff
@@ -19,59 +27,41 @@
-	}
+	@Id
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
+	protected long id; //can this be set to a default?
```
No. :)

haxwell commented:
+1

This is very nice, Jim.. Thanks for your effort... Long term, I think that getText() method in AbstractEntity will be removed and reworked.. but for now, we can leave it.. 

Commented out code, if you could delete that.. 

Other than that, I think its good..

jdisser commented:
Thanks Johnathan.  I knew it was going to be one of the bigger issues but I wanted to learn more about the app and about Java which I did. There is a lot going on in this app.

Here are the final changes with comments removed.
